packages/libs/error-stack/src/fmt/mod.rs

```diff
@@ -0,0 +1,1104 @@
+//!     .attach(2u64)
+//!     .attach("This is going to be at the end".to_owned())
+//!     .attach(3u32)
+//!     .attach(3u32)
+//!     .attach(4u16);
+//!
+//! # owo_colors::set_override(true);
+//! # fn render(value: String) -> String {
+//! #     let backtrace = regex::Regex::new(r"Backtrace No\. (\d+)\n(?:  .*\n)*  .*").unwrap();
+//! #     let backtrace_info = regex::Regex::new(r"backtrace with (\d+) frames \((\d+)\)").unwrap();
+//! #
+//! #     let value = backtrace.replace_all(&value, "Backtrace No. $1\n  [redacted]");
+//! #     let value = backtrace_info.replace_all(value.as_ref(), "backtrace with [n] frames ($2)");
+//! #
+//! #     ansi_to_html::convert_escaped(value.as_ref()).unwrap()
+//! # }
+//! #
+//! # expect_test::expect_file![concat!(env!("CARGO_MANIFEST_DIR"), "/snapshots/fmt__doc.snap")].assert_eq(&render(format!("{report:?}")));
+//! #
+//! # stringify!(
+//! println!("{report:?}");
+//! # );
+//!
+//! # expect_test::expect_file![concat!(env!("CARGO_MANIFEST_DIR"), "/snapshots/fmt_doc_alt.snap")].assert_eq(&render(format!("{report:#?}")));
+//! #
+//! # stringify!(
+//! println!("{report:#?}");
+//! # );
+//! ```
+//! ### `println!("{report:?}")`
+//!
+//! <pre>
+#![doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/snapshots/fmt__doc.snap"))]
+//! </pre>
+//!
+//! ### `println!("{report:#?}")`
+//!
+//! <pre>
+#![doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/snapshots/fmt_doc_alt.snap"))]
+//! </pre>
+//!
+//!
+//! [`Display`]: core::fmt::Display
+//! [`Debug`]: core::fmt::Debug
+//! [`Mutex`]: std::sync::Mutex
+//! [`RwLock`]: std::sync::RwLock
+//! [`Backtrace`]: std::backtrace::Backtrace
+//! [`SpanTrace`]: tracing_error::SpanTrace
+//! [`error_stack::fmt::builtin`]: crate::fmt::builtin
+
+mod hook;
+#[cfg(all(nightly, feature = "experimental"))]
+mod nightly;
+
+use alloc::{
+    borrow::ToOwned,
+    collections::VecDeque,
+    format,
+    string::{String, ToString},
+    vec,
+    vec::Vec,
+};
+use core::{
+    fmt,
+    fmt::{Debug, Display, Formatter},
+    mem,
+};
+
+#[cfg(feature = "std")]
+pub use hook::HookContext;
+use hook::HookContextImpl;
+#[cfg(feature = "std")]
+pub(crate) use hook::Hooks;
+#[cfg(feature = "std")]
+pub use hook::{builtin, Call};
+#[cfg(not(feature = "std"))]
+pub(crate) use hook::{builtin, Call};
+#[cfg(all(nightly, feature = "experimental"))]
+pub use nightly::DebugDiagnostic;
+#[cfg(feature = "glyph")]
+use owo_colors::{OwoColorize, Stream::Stdout, Style as OwOStyle};
+
+use crate::{AttachmentKind, Context, Frame, FrameKind, Report};
+
+/// Modify the behaviour, with which `Line`s returned from hook invocations are rendered.
+///
+/// Lines can either be emitted immediately as the [`Emit::Next`] line, or defer the via
+/// [`Emit::Defer`] until the end of the current "group".
+/// [`Emit::Defer`] does not modify the order of frames, only emits them after all [`Emit::Next`]
+/// frames in the order they were added in.
```

Comment:
We do!

TL;DR: This ensures that backtrace and span trace are always at the end.

We have the following stack of frames:

```
Attachment A
Attachment B
Spantrace
Backtrace
Context A
Attachment C
Attachment  D
Spantrace
Backtrace
Context B
```

we then pull up all contexts:

```
Context A
	Attachment A
	Attachment B
	Spantrace
	Backtrace
Context B
	Attachment C
	Attachment D
	Spantrace
	Backtrace
```

then reverse all attachments:

```
Context A
	Backtrace
	Spantrace
	Attachment B
	Attachment A
Context B
	Backtrace
	Spantrace
	Attachment C
	Attachment D
```

then we evaluate everything

```
Emit::next()
	Emit::defer() <- builtin
	Emit::defer() <- builtin
	Emit::next()
	Emit::next()

...
```

which then reorders to:

```
Emit::next()
	Emit::next()
	Emit::next()
	Emit::defer() <- builtin
	Emit::defer() <- builtin

...
```

leaving backtrace and span trace always at the end

Reply:
Thanks for clarifying. How does it behave with other `Defer`s (user defined), will they appear in the same order they were attached (like other attachments)?
Sorry if this is written down somewhere. 🤞

Reply:
They will go *below* the backtrace, tho I am thinking of changing the order (once again) so that we can ensure that backtrace and span trace are always last.

```
# Reordered Stack

Emit::next()
	A: Emit::defer() <- builtin
	B: Emit::defer() <- builtin
	Emit::next()
	Emit::next()
	C: Emit::defer()

# Currently:

Emit::next()
	Emit::next()
	Emit::next()
	A: Emit::defer() <- builtin
	B: Emit::defer() <- builtin
	C: Emit::defer()

# Potentially:

Emit::next()
	Emit::next()
	Emit::next()
	C: Emit::defer()
	B: Emit::defer() <- builtin
	A: Emit::defer() <- builtin
```